Add Delete functionality for Peek

daverayment opened the pull request:
## Summary of the Pull Request
This PR adds the ability for users to delete the file currently being previewed in Peek.

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [x] **Closes:** #22584
- [x] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [ ] **Tests:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [x] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: [#5168](https://github.com/MicrosoftDocs/windows-dev-docs/pull/5168)

## Detailed Description of the Pull Request / Additional comments
There are several new and changed elements in this PR:

- Navigation has been updated to track forward and backward movement through files. This is necessary so deleting a file always shows the next item in the natural sequence, i.e. the previous item if the user was moving backwards. This replicates Windows Photo and other media viewers.
- The indexes of deleted items are recorded rather than altering the `Items` collection (which could be very large). There are now `DisplayIndex` and `DisplayItemCount` properties, which reflect the calculated position and total. The old `CurrentIndex` still points to the actual index of the current item in the collection, but this is now internal only.
- The MainWindow ViewModel includes quite a bit of navigation-related code. We may wish to split this out into a separate model or provider in the future.
- When previewing a selection of items (where the number and total are shown in the title bar), these counts will always be shown. Previously, this info would not show if the count was 1. (This was likely because opening Peek while a single item is selected expands the Items collection to cover the entire folder and turns off the "(position/total)" display.)
- A text message has been added for when there are no more files remaining, i.e. the user has deleted the last one from their selection (or from the current folder if Peek wasn't opened via multi-selection). It may be worthwhile to add a simple icon by the side here - perhaps an image placeholder graphic? - but I've gone with the simple approach for now.
- Changes were made to the title bar code to ensure that the last filename and "open in application" details were not shown when there are no more files to preview. This was because there was an (entirely reasonable!) assumption in the code that there would always be one or more files being previewed.
- The delete operation sends the item to the Recycle Bin only. It was considered whether pressing `Shift + Delete` could do a permanent delete, and this is certainly possible in the future, but this release errs on the side of caution, especially as it's a new feature and involves deleting things.
- The Delete function itself relies upon a shell file operation, reflecting the fact that we're previewing shell items. This required adding some code to `NativeMethods.cs` for the API call, struct, flags, and error messages.
- Only *file* deletion is allowed. I considered allowing for empty folder deletes, too, but, again, this first release is deliberately conservative.
- Updated Peek documentation.

## Screenshots

### No More Files message
![Screenshot 2024-10-13 230651 (Small)](https://github.com/user-attachments/assets/3692460a-a209-4e57-9379-2d810d6e7d85)

### File Count For Single Remaining File
![image](https://github.com/user-attachments/assets/1e92e30f-bf45-45fd-8927-f119b1b29336)


## Validation Steps Performed

(Manual tests only.)

Tested:
- Single and multi-selection methods of opening Peek.
- Deleting from beginning, end and middle of a sequence (after opening Peek via multi-selection).
- Deleting until no files remained.
- Previewing a folder containing a mix of files and folders.
- Previewing a folder containing a mix of files which can be previewed and unsupported file types.
- Attempting to delete folders via Peek.
- Deleting a file in the selected list via Explorer before it could be previewed.
- Deleting a file from an external USB drive which doesn't support the Recycle Bin, to confirm that the system confirmation dialog is displayed which warns against permanent file deletion.
- Cancelling a permanent delete, confirming the file is re-added to the current list.
- Changing the "Don't show this warning again" setting in the delete confirmation dialog updates the settings.
- Updating the "Ask for confirmation before deleting files" setting in the Settings app is reflected in Peek immediately, and vice-versa. (The Settings app requires navigating to another page then returning to the Peek settings.)

htcfreek commented on src/modules/peek/Peek.FilePreviewer/FilePreview.xaml:
```diff
@@ -102,6 +102,16 @@
+            HorizontalAlignment="Center"
+            VerticalAlignment="Center"
+            Foreground="{StaticResource TextFillColorSecondaryBrush}"
+            Visibility="Collapsed"
```
Binding missing?

daverayment replied:
I removed the Text binding from the original commit because the text property is now being loaded directly from the Resources via the Uid. There is an entry in Resources for `NoMoreFiles.Text`.

htcfreek replied:
I mean the visibility property binding

daverayment replied:
Ah, I understand. The visibility of the TextBlock is controlled directly via the `OnItemPropertyChanged` handler in `FilePreview.xaml.cs`, line 164:

```            NoMoreFiles.Visibility = NumberOfFiles == 0 ? Visibility.Visible : Visibility.Collapsed;```

htcfreek commented on src/modules/peek/Peek.UI/MainWindowViewModel.cs:
```diff
@@ -63,50 +126,124 @@ public void Initialize(HWND foregroundWindowHandle)
+        public void DeleteItem()
         {
-            if (NavigationThrottleTimer.IsEnabled)
+            if (CurrentItem == null || !IsFilePath(CurrentItem.Path))
```
Is this a valid syntax or an exists on disk check? Checking existence on disk is imo a must have. Especially for file on network drives.

daverayment replied:
The `IsFilePath` method checks to ensure the path isn't a folder.

I don't check to see if the file is on a local disk at the moment. Is that what you think should be the case? So we wouldn't allow deletions from a network drive?

htcfreek replied:
I mean checking if the file exists in general. (Is not removed by another process.)

htcfreek replied:
After rethinking I think as we not spam user with screen messages and won't crash an extra check might be obsolete. What do you think?

daverayment replied:
Yes, I think this would be covered by your proposed change to show a message when the delete operation failed. The user would see something like "Delete failed. File not found". We can therefore avoid doing a pre-delete check for whether the file exists.

htcfreek commented on src/modules/peek/Peek.UI/MainWindowViewModel.cs:
```diff
@@ -63,50 +126,124 @@ public void Initialize(HWND foregroundWindowHandle)
+            {
+                wFunc = FO_DELETE,
+                pFrom = path + "\0\0",
+                fFlags = (ushort)(FOF_NOCONFIRMATION | (permanent ? 0 : FOF_ALLOWUNDO)),
```
Does this override the recycle bin property? It is possible to configure the delete confirmation in the recycle bin properties.

How does Fotos app handle it?

daverayment replied:
This replicates how the Photos application handles deletes. Pressing Delete immediately sends the file to the Recycle Bin without popping up a message, and then it moves on to the next file.

The `permanent` argument is there for possible future use, but is always false for now.

htcfreek commented on src/modules/peek/Peek.UI/MainWindowViewModel.cs:
```diff
@@ -63,50 +126,124 @@ public void Initialize(HWND foregroundWindowHandle)
+
+            int result = SHFileOperation(ref fileOp);
+
+            if (result != 0)
```
We need a message (translated) for the user in peek ui. (Imagin youndelete files, nothing hapoens and you don't know why.) We can not expect that users checking thenlog file.

Suggestion: Infobar or teaching tip with severity error.

daverayment replied:
I was in two minds about this while developing. It was a deliberate decision in the end to have the delete operation as a 'best effort' attempt which occurs in the background after navigating away from the current file. If the delete doesn't succeed, I didn't want to pop up a message saying anything technical.

However, I agree that there's the potential for this to be confusing, as the file will have been removed from the preview sequence, but still be on disk.

As you say, maybe the solution is to show a non-modal message.

If the delete fails, do you think the file should be reinstated in the preview sequence so it can be previewed again?

htcfreek replied:
Not readding it is confusing as an existing file isn't previewable anymore until reopening peek.

htcfreek commented:
@daverayment 
Please fix the merge conflicts. (This is the reason for the failing test.)

daverayment commented:
@htcfreek 

Thanks for your comments and suggestions.

I've updated the code with an error InfoBar, which is shown when a delete fails:

![image](https://github.com/user-attachments/assets/a19bda28-7414-4b4d-a5c7-b4799845efc9)

I've added user-friendly messages for the most common failures. Others result in a generic message being shown on the InfoBar.

The failure is also logged, with the code and a more technical error message (taken from `winerror.h`).

If the file still exists after the deletion attempt, it is kept in the items collection for previewing; otherwise it is removed.

An aside: Windows Photo handles files being renamed or deleted differently to Peek, in that it seems to constantly monitor the folder. (Try changing the filename of a photo you're previewing and you can see it update in the UI after a second or two.) We may wish to replicate this functionality in the future, as it would be increase robustness, but that it outside of the scope of this issue.

jaimecbernardo reviewed (changes requested):
Thank you for opening the pull request.
Generally looks good to me, but there's the issue that this might be unintentionally pressed by the user, getting a file deleted he might not have intended. Even though it ends up in the Recycling Bin, I think having a Message Box Prompt first would be safer. I'm OK with a setting to disable this warning, but it should show a confirmation prompt by default.
Hope this makes sense.

htcfreek commented:
> > Even though it ends up in the Recycling Bin, I think having a Message Box Prompt first would be safer.
> 
> Agree. Could also try to get/use the setting for the Recycle Bin itself.

Doesn't the code does this yet? Or better asked: Isn't this handled by Recycle Bin itself?

Jay-o-Way commented:
> Doesn't the code does this yet? Or better asked: Isn't this handled by Recycle Bin itself?

Nope, the code uses `DeleteFile(item)` -> `FO_DELETE` and that's an explicit action. I'd say the setting is an Explorer setting.

daverayment commented:
@jaimecbernardo @htcfreek @Jay-o-Way 

Thank you for the review and guidance. However, I disagree with displaying a prompt.

The default Windows functionality for file deletions is not to show a confirmation prompt. The whole purpose of the Recycle Bin is to to protect against accidental deletions. The behaviour in Peek was deliberately chosen to be consistent with both Windows Explorer and other file viewer tools such as the Windows Photo App and third party apps like IrfanView. I believe users would find a confirmation dialog unnecessary or redundant every time they want to delete a file, which could be a common action when sorting through a large folder of files. Also, PowerToys is supposed to be for power users, so I don't think we should go out of our way to provide guardrails for easily-reversible common mistakes such as accidentally pressing the delete key (which is also not near the default Peek navigation keys). 

To summarise, I don't think we should provide non-default behaviour which would slow down a user's workflow.

jaimecbernardo commented:
I've tried the Windows "Photos" app, it does show a confirmation prompt: 
![image](https://github.com/user-attachments/assets/daef56a2-421a-4c9c-afc5-7a905614df2f)

jaimecbernardo commented:
I imagine you might've already clicked that "Don't show this warning again" @daverayment .
What I'm proposing is having a prompt such as this one from the Windows "Photos" app. But also a setting to allow the prompt to not be shown, but I think the prompt should be shown by default.

jaimecbernardo commented:
I've also just tried the feature on an external drive. As expected, files were permanently deleted without going to the recycle bin. I don't feel easy causing potential data loss with a feature that wasn't there before without some initial guardrails so the user understands that feature is there 😅 Hope this makes sense to you.

daverayment commented:
Thanks, @jaimecbernardo 

Sorry, yes, you're right! I must have clicked the "Don't Show" checkbox when I initially started using Windows Photo. Please forgive the lapse in memory - it was a _long_ time ago... 😬

My main concern was that presenting dialogs every time a delete was requested would get in the way, but I'm fine with your suggestion of replicating the Photo app's behaviour with the "Don't Show" checkbox. This will give me a chance to have a play with the Settings code, too, which I've been avoiding.

I'm very concerned that the delete operation skipped the Recycle Bin for a file on your external device. Would you mind seeing what your Recycle Bin properties are for that drive, please? It is possible to turn off the Recycle Bin on a per-drive basis:

![image](https://github.com/user-attachments/assets/f8b2edf8-ebdc-4d9b-8738-a1aa6ec91dfd)

The Peek code just calls the underlying Shell API, so it should be consistent with doing a delete from Explorer (without holding Shift down); it is a bug if it behaves differently for external drives.

Jay-o-Way commented:
@daverayment `DeleteFile(item)` -> `FO_DELETE` is very explicit. Code often has a "FileRecycle" function too.

jaimecbernardo commented:
> I'm very concerned that the delete operation skipped the Recycle Bin for a file on your external device. Would you mind seeing what your Recycle Bin properties are for that drive, please? It is possible to turn off the Recycle Bin on a per-drive basis:
> 
> ![image](https://private-user-images.githubusercontent.com/7127766/391823855-f8b2edf8-ebdc-4d9b-8738-a1aa6ec91dfd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMyMTkxMTgsIm5iZiI6MTczMzIxODgxOCwicGF0aCI6Ii83MTI3NzY2LzM5MTgyMzg1NS1mOGIyZWRmOC1lYmRjLTRkOWItODczOC1hMWFhNmVjOTFkZmQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMDNUMDk0MDE4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDQ1ODRkZjUwMzhjZDUzZTQxZTY2YWU5ODFiNzE5YjgwYmY5YzVjNGUzZWUxNGQzNTc3NWMyOGY5OGNiOWQwYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.xQvVPKww3kH2RMXiyvuUj-_Tx_rclyPgI1FB_d6fWKQ)

@daverayment  There's nothing on Recycle Bin for this drive and another I've tested. I'm guessing because their file system is FAT32.

htcfreek commented:
@daverayment , @jaimecbernardo 
For best user behavior the setting should be controllable from the dialogue in Peek (disable message) and the settings page (disable and enable) in PT Settings. Is this possible?

daverayment commented:
> @daverayment using a `delete` function in most code languages (like `DeleteFile(item)` -> `FO_DELETE`) is very explicit - it doesn't care about a recycling bin. Code languages often have a "FileRecycle" function too, but...
> 
> > Sadly, C# doesn't have a native API to delete files and folders to Recycle Bin. But Visual Basic has.
> 


Hi, @Jay-o-Way !

Thanks for the message. It is certainly a shame that System.IO doesn't include recycling options when deleting files.

As you noticed, this update to Peek uses the Shell File Operation `FO_DELETE`. This is also used internally by `Microsoft.VisualBasic.FileIO.FileSystem.DeleteFile()` (see [the VB Core library disassembly here](https://source.dot.net/#Microsoft.VisualBasic.Core/Microsoft/VisualBasic/FileIO/FileSystem.vb,1664)). `FO_DELETE` does support a flag to control whether it attempts to move the file to the Recycle Bin (`FOF_ALLOWUNDO`) and various others to control which progress and error messages are shown.

This isn't enough for our needs, though, so I will add the dialog box to prompt the user when they try to delete a file, as previously discussed. Giving users the option to permanently dismiss this warning replicates what the Windows Photo App does, and is a good compromise for power users.

The reason @jaimecbernardo experienced the issue with the FAT32 volume is because I needed to include a flag called `FOF_WANTNUKEWARNING`, which is very well named! I will correct that in the PR. Thank you for the catch, Jaime!

The VB Core library containing `DeleteFile()` is 1.2 MB, so I'm keen to just call the Shell API directly to reduce the potential bloat.

daverayment commented:
> @daverayment , @jaimecbernardo For best user behavior the setting should be controllable from the dialogue in Peek (disable message) and the settings page (disable and enable) in PT Settings. Is this possible?

@htcfreek

Yes, that's what I assumed we'd be doing. I haven't used the Settings system before, but I'll give it a go and I'll ask if I get stuck!

htcfreek commented:
@daverayment 
Are you working on the outstanding tasks (delete message with setting, using recycle bin, fat32 bug)?

/needinfo

daverayment commented:
@htcfreek Yes, I'm still working on this. Sorry about the delay.

htcfreek commented:
@daverayment 
No problem. Only wondering about the `0.88` lable. I removed it now.

daverayment commented:
I've updated the PR to include the Delete confirmation dialog and the setting to control whether it is shown or not.

The dialog looks like this:

![image](https://github.com/user-attachments/assets/e4f7b840-7872-4188-9261-333efbf255fc)

And the new setting:

![image](https://github.com/user-attachments/assets/79d531d2-e218-4b45-845e-8db97986fcb2)

I had an issue during a rebase, so I apologise for the force-push.

A note about one of the implementation details: there is an issue with the focus rectangle being shown on `ContentDialog`s when they are shown in response to a keyboard input, but _not_ a button press or other click/select event. This affects us here because we show the confirmation dialog in response to <kbd>Delete</kbd> being pressed. For a workaround, I settled on zeroing the focus thickness before showing the dialog and restoring it after the user begins interacting with the controls via the keyboard. This is described in more detail here: https://github.com/microsoft/microsoft-ui-xaml/issues/10315. If someone finds a better fix or objects to this approach, please let me know.

I've updated the Peek documentation to describe the new functionality, the dialog and the settings.

daverayment commented:
In further testing I have found a couple of issues with the new dialog, so please don't merge for the time being. The problem can be replicated by following these steps:

1. Open an image for preview in Peek
2. Press <kbd>Delete</kbd>. This shows the new confirmation dialog
3. Close down the main Peek window while the delete confirmation dialog is still being displayed
4. Open Peek again, but for a different image

The confirmation dialog is still displayed, but now refers to the _new_ image.

This is because the dialog is not removed when Peek's MainWindow is hidden, and is bound to the current item being displayed. This would lead to user confusion, so I will fix and re-test.

daverayment commented:
I have updated the delete confirmation dialog to be simpler:

![image](https://github.com/user-attachments/assets/006bf46d-adb5-4796-b818-8e114b814361)

There was no need to include the filename, as it is obvious which file is current.

This revision includes changes so the system permanent deletion warnings are correctly shown:

![image](https://github.com/user-attachments/assets/721636bb-2b4b-4f6d-a153-151c625fcadc)

(The flag **FOF_NO_CONFIRMATION** must not be included for this to work, even though the documentation suggests otherwise.)

I have removed the workaround for the focus issue on the delete confirmation dialog because:

- The Windows Photo app shows the same inconsistency
- It may not be a bug, but a deliberate WinUI design decision when the keyboard is being used to initiate an action. I shouldn't include a workaround until the issue is confirmed

daverayment commented on src/modules/peek/Peek.UI/MainWindowViewModel.cs:
```diff
@@ -3,26 +3,61 @@
+        /// <summary>
+        /// Holds the indexes of each <see cref="IFileSystemItem"/> the user has deleted.
+        /// </summary>
+        private readonly HashSet<int> _deletedItemIndexes = [];
+
```
The use of collection expressions in C# 12 and above is valid and recommended, so I'm going to just set this to resolved and downvote the suggestion.

drawbyperpetual commented on src/modules/peek/Peek.UI/Native/NativeMethods.cs:
```diff
@@ -51,5 +51,85 @@ public enum AssocStr
+        /// Common error codes when calling SHFileOperation to delete a file.
+        /// </summary>
+        /// <remarks>See winerror.h for full list.</remarks>
+        public static readonly Dictionary<int, string> DeleteFileErrors = new()
```
As far as I can see, these are not localized and are shown directly to the user. This needs to be addressed.

drawbyperpetual commented:
Any reason not to include an explicit button for this action? The Windows Photos app has one and the other Peek action that one can perform with files (open in default app) also has one.

drawbyperpetual commented on src/settings-ui/Settings.UI/ViewModels/PeekViewModel.cs:
I notice that when the Peek page is open on the Settings page and the `ConfirmFileDelete` setting is changed via Peek itself, the toggle does not update on the Settings page. Other Settings pages that allow settings to be changed externally usually support this by registering a file-system watcher on the settings file. Would be nice to have this for Peek too now.

htcfreek replied:
Small side note: For PowerToys Run the file watcher seems to not work. The only page that works is Awake.

daverayment replied:
@drawbyperpetual Yes, there is a watcher in the Peek application itself to respond to Settings app changes, but not the other way around.

Would you like me to raise this as an issue for Peek, or do you regard it as a blocker for this PR?

On a more general level, I think it may be worth reviewing each application to gather information about which have fully synchronised settings, and which require the user to reopen either the application or the Settings app page to refresh the changes. I recently raised an issue for Image Resizer which has the opposite issue to Peek, where Settings app changes weren't synchronised to the running application.

drawbyperpetual replied:
It would be nice if it were done in this PR. The reason I say that is that it is this PR that is introducing the first (and to date, only) setting to Peek that can be modified from outside the Settings app.

daverayment replied:
Sure. I'll have a look.

daverayment replied:
I've pushed an update to handle the synchronisation of the setting from Peek back to the Settings app.

drawbyperpetual commented on src/modules/peek/Peek.UI/PeekXAML/MainWindow.xaml.cs:
```diff
@@ -56,6 +63,53 @@ public MainWindow()
             AppWindow.Closing += AppWindow_Closing;
         }
 
+        private async void Content_KeyUp(object sender, KeyRoutedEventArgs e)
```
I have a strange issue where, when the confirmation dialog is disabled, I notice Peek makes multiple spurious delete attempts. See in the example below where `2.dat` is selected and the delete key is pressed once. Peek deletes `2.dat`, then switches to `1.dat`, deletes *that* file, then finally tries to delete it again (which fails)!

It could be that multiple `KeyUp` events are raised for a single key-press, in which case a solution [such as this](https://github.com/microsoft/PowerToys/blob/22e29d125332e972987c55e24c561a0c992a81df/src/modules/AdvancedPaste/AdvancedPaste/AdvancedPasteXAML/Pages/MainPage.xaml.cs#L151) may be appropriate. In any case, if this is an issue with this PR and not with my machine, this is a serious bug that should be addressed.

![peek_multiple_delete](https://github.com/user-attachments/assets/54fa8214-f12f-4ad4-89c9-6de3723213a9)

daverayment replied:
This looks like an issue caused by multiple `KeyUp` events being triggered because I hadn't set `Handled` to true in the event handler. The event will keep propagating and it looks like it will retrigger the same handler under certain circumstances. I had trouble reproducing it on my laptop, but I have just seen double-deletes on my other machine.

Could you please try changing the `Content_KeyUp` code in **MainWindow.xaml.cs** to:

``` csharp
    private async void Content_KeyUp(object sender, KeyRoutedEventArgs e)
    {
        if (e.Key == Windows.System.VirtualKey.Delete)
        {
            e.Handled = true;
            await DeleteItem();
        }
    }
```

and retrying.

There is already a throttle in for navigation events (100ms) and a 200ms delay for the delete operation itself. `KeyUp` events don't repeat, so holding down <kbd>delete</kbd> should be OK.

drawbyperpetual replied:
Thanks; that suggested change indeed fixes the issue for me. You can commit that.

daverayment replied:
Done. Thanks again for the review.

daverayment commented:
> Any reason not to include an explicit button for this action? The Windows Photos app has one and the other Peek action that one can perform with files (open in default app) also has one. Another reason to include an explicit button is that the Delete key does not seem to work well with text files because of [this issue](https://github.com/microsoft/PowerToys/issues/28715)

Thanks for your review, @drawbyperpetual 😊

The intention was to keep the UI minimal, similar to why there are no icons for basic navigation. I also wasn't keen on having a destructive action button in the title bar. However, I do realise that's personal preference.

With regard to the WebView-related focus, the current behaviour is a good thing with regard to deletes. There is a difference in user expectation between pressing <kbd>escape</kbd> (should exit the viewer) and pressing <kbd>delete</kbd> while viewing a text file. When viewing text, Peek gives the user a caret and allows them to select text, so it is reasonable to assume that some of them will try to edit the file contents, including pressing <kbd>delete</kbd> to remove a paragraph or whatever. These delete keystrokes shouldn't bubble up to the KeyUp handler.

In any case, I'm happy to add the button or wait to see if there is demand for it from the community after the functionality is released. Please let me know.

drawbyperpetual commented:
> > Any reason not to include an explicit button for this action? The Windows Photos app has one and the other Peek action that one can perform with files (open in default app) also has one. Another reason to include an explicit button is that the Delete key does not seem to work well with text files because of [this issue](https://github.com/microsoft/PowerToys/issues/28715)
> 
> Thanks for your review, @drawbyperpetual 😊
> 
> The intention was to keep the UI minimal, similar to why there are no icons for basic navigation. I also wasn't keen on having a destructive action button in the title bar. However, I do realise that's personal preference.
> 
> With regard to the WebView-related focus, the current behaviour is a good thing with regard to deletes. There is a difference in user expectation between pressing escape (should exit the viewer) and pressing delete while viewing a text file. When viewing text, Peek gives the user a caret and allows them to select text, so it is reasonable to assume that some of them will try to edit the file contents, including pressing delete to remove a paragraph or whatever. These delete keystrokes shouldn't bubble up to the KeyUp handler.
> 
> In any case, I'm happy to add the button or wait to see if there is demand for it from the community after the functionality is released. Please let me know.

Sure; it can always be added later.

drawbyperpetual commented on src/modules/peek/Peek.UI/MainWindowViewModel.cs:
```diff
@@ -3,26 +3,61 @@
+using System.Globalization;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
```
Nit: Blank line after `System` namespaces seems to be a convention now for some reason.

drawbyperpetual commented on src/modules/peek/Peek.UI/MainWindowViewModel.cs:
```diff
@@ -63,50 +136,216 @@ public void Initialize(HWND foregroundWindowHandle)
-            NavigationThrottleTimer.Start();
+            bool skipConfirmation = skipConfirmationChecked ?? false;
+            bool shouldShowConfirmation = !skipConfirmation;
+            Application.Current.GetService<IUserSettings>().ConfirmFileDelete = shouldShowConfirmation;
```
Nit: Shouldn't this dependency be added via constructor injection?

daverayment replied:
Absolutely. I have a refactor of most of the navigation-related code, which is incoming. The View Model is currently closely coupled to things like the throttling behaviour, the way files are navigated and so on, making it impossible to test the individual components.

I will handle these updates via a separate PR, as it changes quite a few things around.

drawbyperpetual commented on src/modules/peek/Peek.UI/MainWindowViewModel.cs:
```diff
@@ -63,50 +136,216 @@ public void Initialize(HWND foregroundWindowHandle)
+                    if (Path.GetPathRoot(item.Path) is string root)
+                    {
+                        var driveInfo = new DriveInfo(root);
+                        Logger.LogInfo($"User cancelled deletion of \"{item.Name}\" on " +
```
Filenames are private user info; we shouldn't be logging them

daverayment replied:
I've made this line generic now in a new update.

drawbyperpetual commented on src/modules/peek/Peek.UI/PeekXAML/MainWindow.xaml.cs:
```diff
@@ -3,12 +3,13 @@
 
 using System;
-
+using System.Threading.Tasks;
```
Nit: Blank line (same as elsewhere)

daverayment replied:
Updated in new commit.

drawbyperpetual reviewed (approved):
Code looks good and works well in my testing!

jaimecbernardo commented:
/azp run

jaimecbernardo commented:
/azp run
